src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java

```diff
@@ -0,0 +1,184 @@
+        return referenceBook;
+    }
+
+    public static List<Text> getOverlayText() {
+        if (CodeClient.location instanceof Dev dev) {
+            if (!Config.getConfig().ActionViewer) return null;
+            ItemStack item;
+            if (action == null) {
+                if (!reference) return null;
+                if (book == null) {
+                    item = getReferenceBook();
+                    if (item == null) return null;
+                    var displayBook = item.copy();
+                    var name = displayBook.getName().copy().withColor(0xFFFFFF);
+                    displayBook.setCustomName(name); // continuity with actions
+                    book = displayBook;
+                } else {
+                    item = book;
+                }
+            } else {
+                item = action.icon.getItem();
+            }
+            return item.getTooltip(null, TooltipContext.BASIC);
+        }
+        return null;
+    }
+
+    public static void render(DrawContext context, HandledScreen<?> handledScreen) {
+        if (!(handledScreen instanceof GenericContainerScreen || handledScreen instanceof CustomChestMenu))
+            return;
+        var text = getOverlayText();
+        if (text == null) return;
+
+        drawTooltip(context, handledScreen, transformText(text), 176, 0, 300);
+    }
+
+    private static List<TooltipComponent> transformText(List<Text> texts) {
+        var orderedText = Lists.transform(texts, Text::asOrderedText);
+        return  orderedText.stream().map(TooltipComponent::of).toList();
+    }
+
+    // this implementation of draw tooltip allows a change in the z-index of the rendered tooltip.
+    private static void drawTooltip(DrawContext context, HandledScreen<?> handledScreen, List<TooltipComponent> components, int x, int y, int z) {
+        var textRenderer = CodeClient.MC.textRenderer;
+
+        var positioner = ActionTooltipPositioner.INSTANCE;
+        if (handledScreen instanceof CustomChestMenu menu) {
+            var handler = menu.getScreenHandler();
+            positioner.setBackgroundHeight(handler.numbers.MENU_HEIGHT);
+            x += handler.numbers.MENU_WIDTH - 176;
+        } else {
+            positioner.setBackgroundHeight(166);
+        }
+
+        if (components.isEmpty()) return;
+        int tooltipWidth = 0;
+        int tooltipHeight = components.size() == 1 ? -2 : 0;
+
+        TooltipComponent tooltipComponent;
+        for (Iterator<TooltipComponent> iterator = components.iterator(); iterator.hasNext(); tooltipHeight += tooltipComponent.getHeight()) {
+            tooltipComponent = iterator.next();
+            int width = tooltipComponent.getWidth(textRenderer);
+            if (width > tooltipWidth) {
+                tooltipWidth = width;
+            }
+        }
+        Vector2ic vector = positioner.getPosition(context.getScaledWindowWidth(), context.getScaledWindowHeight(), x, y, tooltipWidth, tooltipHeight);
```

Comment:
if it's going of the edge, such as a small screen size, we should probably make hovering over the tooltip (or where it normally) is move it into the screen

Reply:
going to look into this, also do you have any suggestions on how to handle tool tips that are too tall to display on screen? originally i was thinking of doing something with scaling but i couldn't think of a way to do that.